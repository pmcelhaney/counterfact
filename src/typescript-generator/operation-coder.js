/* eslint-disable no-magic-numbers */
import nodePath from "node:path";

import { Coder } from "./coder.js";
import { OperationTypeCoder } from "./operation-type-coder.js";

export class OperationCoder extends Coder {
  requestMethod() {
    return this.requirement.url.split("/").at(-1).toUpperCase();
  }

  names() {
    return super.names(this.requestMethod());
  }

  write() {
    const responses = this.requirement.get("responses");

    const [firstStatusCode] = responses.map(
      (response, statusCode) => statusCode,
    );
    const [firstResponse] = responses.map((response) => response.data);

    if (!("content" in firstResponse || "schema" in firstResponse)) {
      return "() => { /* no response content specified in the OpenAPI document */ }";
    }

    return `($) => {
      return $.response[${
        firstStatusCode === "default" ? 200 : firstStatusCode
      }].random();
    }`;
  }

  typeDeclaration(namespace, script) {
    const operationTypeCoder = new OperationTypeCoder(this.requirement);

    return script.importType(operationTypeCoder);
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath
      .join("path", pathString)
      .replaceAll("\\", "/")}.types.ts`;
  }
}
