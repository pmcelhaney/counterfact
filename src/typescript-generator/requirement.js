export class Requirement {
  constructor(data, url = "", specification = undefined) {
    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  has(item) {
    return item in this.data;
  }

  get(item) {
    if (!this.has(item)) {
      return undefined;
    }

    return new Requirement(
      this.data[item],
      `${this.url}/${item}`,
      this.specification,
    );
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification,
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback(this.select(this.escapeJsonPointer(key)), key);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach((value, key) => result.push(callback(value, key)));

    return result;
  }

  flatMap(callback) {
    // eslint-disable-next-line unicorn/prefer-array-flat-map
    return this.map(callback).flat();
  }

  find(callback) {
    // eslint-disable-next-line init-declarations
    let result;

    this.forEach((value, key) => {
      // eslint-disable-next-line  n/callback-return
      if (result === undefined && callback(value, key)) {
        result = value;
      }
    });

    return result;
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
