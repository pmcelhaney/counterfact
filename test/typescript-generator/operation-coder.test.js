import { describe, expect, it } from "@jest/globals";
import prettier from "prettier";

import { OperationCoder } from "../../src/typescript-generator/operation-coder.js";
import { Requirement } from "../../src/typescript-generator/requirement.js";

function format(code) {
  return prettier.format(code, { parser: "typescript" });
}

describe("an OperationCoder", () => {
  it("generates a list of potential names", () => {
    const coder = new OperationCoder(
      new Requirement({}, "#/paths/hello/get"),
      "get",
    );

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual(["GET", "GET2", "GET3"]);
  });

  it("creates a type declaration", () => {
    const coder = new OperationCoder(
      new Requirement({}, "#/paths/hello/get"),
      "get",
    );

    const script = {
      importType() {
        return "HTTP_GET";
      },
    };

    expect(coder.typeDeclaration(undefined, script)).toBe("HTTP_GET");
  });

  it("returns the module path", () => {
    const coder = new OperationCoder(
      new Requirement({}, "#/paths/hello~1world/get", "get"),
      "get",
    );

    expect(coder.modulePath()).toBe("path/hello/world.types.ts");
  });

  it("generates a complex get operation", async () => {
    const requirement = new Requirement(
      {
        parameters: [
          { in: "path", name: "id", schema: { type: "string" } },
          { in: "query", name: "name", schema: { type: "string" } },
        ],

        responses: {
          200: {
            content: {
              "application/json": {
                examples: {
                  "first-example": {
                    value: "first",
                  },

                  "second-example": {
                    value: "second",
                  },
                },

                schema: { $ref: "#/components/schemas/Example" },
              },
            },
          },

          400: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Error" },
              },
            },
          },

          default: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Error" },
              },
            },
          },
        },
      },
      "#/paths/hello/get",
    );

    const coder = new OperationCoder(requirement, "get");

    await expect(
      format(
        coder.write({
          import() {
            return "schema";
          },

          importType() {
            return "Type";
          },
        }),
      ),
    ).resolves.toMatchSnapshot();
  });

  it("generates a simple post operation", async () => {
    const requirement = new Requirement(
      {
        responses: {
          default: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Example" },
              },
            },
          },
        },
      },
      "#/paths/hello/post",
    );

    const coder = new OperationCoder(requirement, "get");

    await expect(
      format(
        coder.write({
          import() {
            return "schema";
          },

          importType() {
            return "Type";
          },
        }),
      ),
    ).resolves.toMatchSnapshot();
  });
});
