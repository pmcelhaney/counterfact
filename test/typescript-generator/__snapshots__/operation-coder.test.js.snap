// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`an OperationCoder generates a complex get operation 1`] = `
"($) => {
  return $.response[200].random();
};
"
`;

exports[`an OperationCoder generates a simple post operation 1`] = `
"($) => {
  return $.response[200].random();
};
"
`;
